### PR TITLE
Add structures for TPM2_GetRandom command.

### PR DESCRIPTION
**Pull request by TRJensen**
- Add test crude test to catch entirely random data.
- Ordered tests into sub modules matching TPM command chapters.
- Fixes https://github.com/tpm-rs/tpm-rs/issues/118

### CODE REVIEW

File: client/tests/simulator_tests.rs [hunk 1]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
 
+#[test]
+fn test_get_random() {
+    const REQUEST_SIZE: u16 = 32;
+    const MIN_RECEIVED: u16 = 32;
+
+    let (_sim_lifeline, mut tpm) = get_started_tpm();
+
+    let command = GetRandomCmd {
+        bytes_requested: REQUEST_SIZE,
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:94]: Personally I'm not sure I see the need for a constant for this, seems to make the test less-readable IMO.

File: client/tests/simulator_tests.rs [hunk 2]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+    };
+
+    // We allow panic in test cases.
+    let resp = run_command(&command, &mut tpm).expect("Failed running command.");
+
+    // Validate that retrieved data contains at least 32 values (SHA256 digest),
+    // and is less that requested amount.
+    assert!(resp.random_bytes.size >= MIN_RECEIVED);
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:102]: Correct me if I'm wrong, but it should only ever be less-than the requested size (in the case where the requested bytes are greater than the size of a `Tpm2bDigest`)?

File: client/tests/simulator_tests.rs [hunk 3]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+    assert!(resp.random_bytes.size <= REQUEST_SIZE);
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:103]: I wonder if we should assert instead <= size of the `Tpm2bDigest`?

File: client/tests/simulator_tests.rs [hunk 4]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+
+    // Lets pull out the actual data as a slice for convenience
+    let random_slice = &resp.random_bytes.buffer[0..resp.random_bytes.size as usize];
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:106]: You don't need to state the first 0 in `[0..*]`, if you leave it off it's implied.

File: client/tests/simulator_tests.rs [hunk 5]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
     assert_eq!(received, expected);
 }
 
+#[test]
+fn test_get_random() {
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:87]: Probably worth having a test that checks that if you ask for a size larger than `Tpm2bDigest` that it cuts back the response properly and doesn't fail or panic.

File: client/tests/simulator_tests.rs [hunk 6]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
     assert_eq!(received, expected);
 }
 
+#[test]
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:86]: Test structure idea: I can see this is going to get unweildy as more commands are implemented. WDYT of each command having a separate module and then you can implement tests specific to commands in those submodules?

Then you can name the tests more specifically what they test and it'll look nicer in test output.

So for instance, `simulator_tests::get_random::trivial_repeat_check` would be this test, and `simulator_tests::get_random::clamps_to_tpm2_digest_size` would be the other test I've asked for below.

File: client/tests/simulator_tests.rs [hunk 7]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+        let mut i = 0;
+        for i in 0..random_slice.len() as usize {
+            // Calculate the circular distance modulo 256 going upwards from first value.
+            // This will give an even probability distribution of values [0 ; 255].
+            let diff: u16 = (256 as u16 + random_slice[i] as u16 - last_value as u16) % 256;
+            count[(diff >> 4) as usize] += 1;
+            last_value = random_slice[i];
+        }
+
+        for i in 0..count.len() as usize {
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:125]: Can be expressed more succinctly:

```rs
let zeros = count.iter().filter(|v| v == 0).count();
```

File: client/tests/simulator_tests.rs [hunk 8]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+    // Print for test inspection.
+    println!("Got random sequence: {:?}", random_slice);
+
+    // Crude randomness test. Confirm we did not get identical data sequences.
+    {
+        let mut zeros: u16 = 0;
+        let mut count: [u16; 16] = [0; 16]; // 16 slots.
+        let mut last_value = random_slice[random_slice.len() - 1];
+        let mut i = 0;
+        for i in 0..random_slice.len() as usize {
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:117]: It might be slightly nicer to use iterators to get your `prev` and `next` instead of relying on indexing the slice:

```rs
for (prev, curr) in random_slice.iter().zip(random_slice.iter().cycle().skip(1)) {
    // Your calculation.
}
```

Though I'm not sure if my brain is just broken from doing too much work with iterators. So take this as an optional. Note that it doesn't technically iterate in the exact same order, but I think it accomplishes the same thing, as it iterates over all pairs of `(prev, curr)` (the difference is the pair `([len()-1],[0])` happens at the end in the iterator construction above).

File: client/tests/simulator_tests.rs [hunk 9]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+
+    // Print for test inspection.
+    println!("Got random sequence: {:?}", random_slice);
+
+    // Crude randomness test. Confirm we did not get identical data sequences.
+    {
+        let mut zeros: u16 = 0;
+        let mut count: [u16; 16] = [0; 16]; // 16 slots.
+        let mut last_value = random_slice[random_slice.len() - 1];
+        let mut i = 0;
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:116]: I wouldn't make this local, not needed.

File: client/tests/simulator_tests.rs [hunk 10]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+    let random_slice = &resp.random_bytes.buffer[0..resp.random_bytes.size as usize];
+
+    // Print for test inspection.
+    println!("Got random sequence: {:?}", random_slice);
+
+    // Crude randomness test. Confirm we did not get identical data sequences.
+    {
+        let mut zeros: u16 = 0;
+        let mut count: [u16; 16] = [0; 16]; // 16 slots.
+        let mut last_value = random_slice[random_slice.len() - 1];
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:115]: `last_value` not needed if you use iterators as detailed below (optional).

File: client/tests/simulator_tests.rs [hunk 11]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+        let mut i = 0;
+        for i in 0..random_slice.len() as usize {
+            // Calculate the circular distance modulo 256 going upwards from first value.
+            // This will give an even probability distribution of values [0 ; 255].
+            let diff: u16 = (256 as u16 + random_slice[i] as u16 - last_value as u16) % 256;
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:120]: I assume this `256` is to avoid underflow, but personally I'd just embrace it and use `u8` for your logic here: https://doc.rust-lang.org/std/primitive.u8.html#method.wrapping_sub

File: client/tests/simulator_tests.rs [hunk 12]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+    // Validate that retrieved data contains at least 32 values (SHA256 digest),
+    // and is less that requested amount.
+    assert!(resp.random_bytes.size >= MIN_RECEIVED);
+    assert!(resp.random_bytes.size <= REQUEST_SIZE);
+
+    // Lets pull out the actual data as a slice for convenience
+    let random_slice = &resp.random_bytes.buffer[0..resp.random_bytes.size as usize];
+
+    // Print for test inspection.
+    println!("Got random sequence: {:?}", random_slice);
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:109]: Can be written as `println!("Got random sequence: {random_slice:?}");`

File: client/tests/simulator_tests.rs [hunk 13]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+            count[(diff >> 4) as usize] += 1;
+            last_value = random_slice[i];
+        }
+
+        for i in 0..count.len() as usize {
+            if count[i] == 0 {
+                zeros += 1;
+            }
+        }
+        println!(
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:130]: Can be written as: 

```rs
        println!(
            "Filled circular distance of RND bytes into {count} slots: {}",
            count.len()
        );
```

File: client/tests/simulator_tests.rs [hunk 14]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
...
+            "Filled circular distance of RND bytes into {} slots: {:?}",
+            count.len(),
+            count
+        );
+        println!(
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:135]: Can be written as:

```rs
        println!(
            "After drawing 32 values, we found {zeros} empty slots"
        );
```

Not sure we need to say how many we allow, the assertion will say that for us, you can add additional text to `assert` by the way, if you'd like to provide the added easy-to-read English context there: https://doc.rust-lang.org/std/macro.assert.html

**Review by trent-reed — COMMENTED** [review]
Reviewed, thanks! Take what makes sense and push back on what doesn't.

File: client/tests/simulator_tests.rs (hunk 5 above)
Comment (oddcoder) [inline client/tests/simulator_tests.rs:87]: I have one comment here, I think this test is overkill, to my understanding we do basic statistical test for randomness. It can be simplified to making sure that the result array is not all zeros.

Also I would like to see more tests like getting zero bytes of randomness, 1 byte of randomness, and trying to get more bytes than the limits.

File: client/tests/simulator_tests.rs (hunk 3 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:103]: Cleaned up

File: client/tests/simulator_tests.rs (hunk 4 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:106]: Fixed 👍

File: client/tests/simulator_tests.rs (hunk 5 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:87]: Fixed 👍

File: client/tests/simulator_tests.rs (hunk 9 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:116]: Fixed 👍

File: client/tests/simulator_tests.rs (hunk 10 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:115]: I have simplified the test - so construct no longer used 👍

File: client/tests/simulator_tests.rs (hunk 11 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:120]: I have simplified the test - so construct no longer used 👍

File: client/tests/simulator_tests.rs (hunk 5 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs]: Changed to test for simple repeated values test - should catch the typical memory dumps like 00 00 00 01 00 00 00 00 or typical memset patterns different from 0.

File: client/Cargo.toml [hunk 15]
```diff
@@ -9,6 +9,8 @@ tpm-simulator-tests = []
 [dependencies]
 tpm2-rs-base = { workspace = true }
 tpm2-rs-marshalable = { workspace = true }
+tpm2-rs-unionify = { workspace = true }
+tracing = { workspace = true }
```
Comment (trent-reed) [inline client/Cargo.toml:13]: :(

Hard disagree.

If you needed this for testing, I would have added it to dev-dependencies.

Comment (trent-reed) [reply client/Cargo.toml:13]: Oh, I thought you already merged this, but it's actually still in review - my bad.

Anyways, I don't think we need this, personally. I don't see what this really adds that we don't get well-enough with just `println` in test situations, but I'd be okay adding it as long as you only add it to `dev-dependencies`

File: client/Cargo.toml [hunk 16]
```diff
@@ -9,6 +9,8 @@ tpm-simulator-tests = []
 [dependencies]
 tpm2-rs-base = { workspace = true }
 tpm2-rs-marshalable = { workspace = true }
+tpm2-rs-unionify = { workspace = true }
```
Comment (trent-reed) [inline client/Cargo.toml:12]: Any particular reason the client is adding this? Probably for sizing in tests. In that case please move to dev-dependencies.

File: client/tests/simulator_tests.rs [hunk 17]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+    };
+
+    // We allow panic in test cases.
+    let resp = run_command(&command, &mut tpm).expect("Failed running command.");
+
+    // Lets pull out the actual data as a slice for convenience
+    let random_slice = &resp.random_bytes.buffer[..resp.random_bytes.size as usize];
+
+    // We should have received less than what the client implementation supports.
+    assert!(random_slice.len() <= TpmtHa::UNION_SIZE);
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:103]: Optional: Instead of having the comments here, you could mix them into the assert so that they print on failure.

e.g. `assert!(..., "We expected to have received no more than the maximum supported digest size")`

File: client/tests/simulator_tests.rs [hunk 18]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+
+    // We should have received minimum size of SHA256 (minimum mandatory digest size).
+    assert!(random_slice.len() >= TPM2_SHA256_DIGEST_SIZE as usize);
+
+    // Print for test inspection.
+    trace!("Got random sequence: {:?}", random_slice);
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:109]: Maybe just me, but I don't think I'd take a crate dependency for debug prints in tests. You only see the prints when the test fails anyways, and I think that's a desirable trait, myself. Why are we adding this `trace` stuff?

File: client/tests/simulator_tests.rs [hunk 19]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+
+    // Duplicate value trap, to catch any same value sequences.
+    {
+        let mut same_twice: u16 = 0;
+        for i in 1..random_slice.len() as usize {
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:114]: Optional: Could be done with iterators; I find this more readable though not all may:

```rs
let same_twice_occurrences = random_slice.iter()
    .zip(random_slice.iter().skip(1))
    .filter(|(prev, next)| prev == next)
    .count();
```

---

Ideally Rust would provide a window function, which is what you really want here, but I don't know that I want to pull in itertools for a single function. Seems likely to someday be in the std so I guess we can just build the window manually for now using `zip`.

File: client/tests/simulator_tests.rs [hunk 20]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+        };
+
+        // We allow panic in test cases.
+        let resp = run_command(&command, &mut tpm).expect("Failed running command.");
+
+        // Lets pull out the actual data as a slice for convenience
+        let random_slice = &resp.random_bytes.buffer[..resp.random_bytes.size as usize];
+
+        // We should have received zero bytes.
+        assert!(random_slice.len() == i);
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:144]: Probably better to `assert_eq!` instead.

Comment (trent-reed) [reply client/tests/simulator_tests.rs:144]: Also, you probably don't need to form the slice, you could probably just inspect `resp.random_bytes.size` for this test.

File: client/tests/simulator_tests.rs [hunk 21]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+        );
+        assert!(same_twice <= random_slice.len() as u16 / 2); // Same value repeated in half of the values unlikely with > 32 values.
+    }
+}
+
+#[test]
+fn test_get_random_small_sizes() {
+    let (_sim_lifeline, mut tpm) = get_started_tpm();
+
+    for i in 0..4 as usize {
```
Comment (trent-reed) [inline client/tests/simulator_tests.rs:132]: Why up to `4`? Just to test the lower-level functionality from `u32` vs `u64` that might be treated specially (ala @oddcoder's ongoing PR)?

In that case maybe worth going up to size of `u64` or `u128` to set some clear obvious boundary defined by a type.

Comment (trent-reed) [inline client/tests/simulator_tests.rs:132]: Unneeded `as usize` cast, I think

**Review by trent-reed — COMMENTED** [review]
Re-reviewed, thanks!

File: client/tests/simulator_tests.rs (hunk 8 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs]: I love this Rust construct 👍 I played a little around with it 🙂 The zip will only provide a value pair iterator corresponding to the shorter of the two iterators. Cycle() loops the one iterator, so it infinitely wraps around. And skip offsets the iterator 1 element in this case. In the simplified crude test, I just need adjacent value pairs, so I adopted this one and skipped the cycle 👍

File: client/Cargo.toml (hunk 15 above)
Comment (TRJensen) [reply client/Cargo.toml:13]: I have moved all texts into the assert line, so that we only get the print in case of a failed test 👍 🙂

File: client/Cargo.toml (hunk 16 above)
Comment (TRJensen) [reply client/Cargo.toml:12]: Moved 👍 🙂

File: client/tests/simulator_tests.rs (hunk 17 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:103]: Fixed.

File: client/tests/simulator_tests.rs [hunk 22]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+    let resp = run_command(&command, &mut tpm).expect("Failed running command.");
+
+    // Lets pull out the actual data as a slice for convenience
+    let random_slice = &resp.random_bytes.buffer[..resp.random_bytes.size as usize];
+
+    // We should have received less than what the client implementation supports.
+    assert!(random_slice.len() <= TpmtHa::UNION_SIZE);
+
+    // We should have received minimum size of SHA256 (minimum mandatory digest size).
+    assert!(random_slice.len() >= TPM2_SHA256_DIGEST_SIZE as usize);
```
[the remark this answers is not in the document]
Comment (TRJensen) [reply client/tests/simulator_tests.rs:106]: Fixed

File: client/tests/simulator_tests.rs (hunk 18 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:109]: Original idea was to add tracing so it can be disabled/enabled at will - but adding messages to the assert messages is fine here. Fixed 👍 🙂

File: client/tests/simulator_tests.rs (hunk 19 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:114]: Ah - another great Rust construct 🙂 I am still thinking too much in the "C" programming strategies. But here the Rust compiler will optimize (i hope) by it self. I have adopted this strategy - thanks again for treat tips 🙂 👍

File: client/tests/simulator_tests.rs [hunk 23]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+            if random_slice[i - 1] == random_slice[i] {
+                same_twice += 1;
+            }
+        }
+
+        trace!(
+            "After {} values, we got duplicate value {same_twice} times.",
+            random_slice.len()
+        );
+        assert!(same_twice <= random_slice.len() as u16 / 2); // Same value repeated in half of the values unlikely with > 32 values.
```
[the remark this answers is not in the document]
Comment (TRJensen) [reply client/tests/simulator_tests.rs:124]: Fixed

File: client/tests/simulator_tests.rs [hunk 24]
```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
...
+            bytes_requested: i as u16,
+        };
+
+        // We allow panic in test cases.
+        let resp = run_command(&command, &mut tpm).expect("Failed running command.");
+
+        // Lets pull out the actual data as a slice for convenience
+        let random_slice = &resp.random_bytes.buffer[..resp.random_bytes.size as usize];
+
+        // We should have received zero bytes.
```
[the remark this answers is not in the document]
Comment (TRJensen) [reply client/tests/simulator_tests.rs:143]: Removed

File: client/tests/simulator_tests.rs (hunk 20 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:144]: Fixed

File: client/tests/simulator_tests.rs (hunk 21 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:132]: Fixed

File: client/tests/simulator_tests.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (TRJensen) [reply client/tests/simulator_tests.rs:114]: Changed to zip() skip() filter() count() construct 🙂 👍 

Fixed

File: client/tests/simulator_tests.rs (hunk 6 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:88]: I think it is a great idea. We already moved all command structures into sub folders. And these tests may explode in future. I will move it once we are done reviewing the current test cases 👍 🙂 Or we can create a separate task for this cleanup, no to have too many changes in one go.

File: client/tests/simulator_tests.rs (hunk 7 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs]: Fixed (used similar in new construct).

File: client/tests/simulator_tests.rs (hunk 13 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs]: Fixed (same idea brought forward to latest changes).

File: client/tests/simulator_tests.rs (hunk 21 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs:132]: I have changed to test just 0 and 1. Zero requests seems to be allowed for some reason. And 1 is the minimal case.
The number of requested bytes is (by TPM spec and our request struct definition) limited to u16. So I modified the max test to handle 2 large requests. (0xFFF and 0xFFFF).

File: base/src/lib.rs [hunk 25]
```diff
@@ -584,49 +584,49 @@ pub type Tpm2bOperand = Tpm2bDigest;
 #[repr(C)]
 #[derive(Clone, Copy, PartialEq, Debug)]
 pub struct Tpm2bData {
-    size: u16,
+    pub size: u16,
```
Comment (oddcoder) [inline base/src/lib.rs:587]: I do not think tpm2b family of types should expose size and buffer content. It might be better to expose that functionality over 3 API.

```
impl AsRef<[u8]> for Tpm2b*

impl AsMut<[u8]> for Tpm2b*
```

File: client/tests/simulator/capability.rs [hunk 26]
```diff
@@ -0,0 +1,42 @@
+#![cfg(feature = "tpm-simulator-tests")]
```
Comment (trent-reed) [inline client/tests/simulator/capability.rs:1]: Unneeded, there's already a conditional compile in the parent module.

File: client/tests/simulator/mod.rs [hunk 27]
```diff
@@ -0,0 +1,9 @@
+#![cfg(feature = "tpm-simulator-tests")]
```
Comment (trent-reed) [inline client/tests/simulator/mod.rs:1]: Unneeded, there's already a conditional compile in the parent module.

File: client/tests/simulator/random.rs [hunk 28]
```diff
@@ -0,0 +1,98 @@
+#![cfg(feature = "tpm-simulator-tests")]
```
Comment (trent-reed) [inline client/tests/simulator/random.rs:1]: Unneeded, there's already a conditional compile in the parent module.

File: client/tests/simulator/mod.rs [hunk 29]
```diff
@@ -0,0 +1,9 @@
...
+
+// Please see "simulator_tests.rs" for details.
```
Comment (trent-reed) [inline client/tests/simulator/mod.rs:3]: Yeah, this is where we get into something I dislike about current rust language - the `tests` directory is kind of magical, in that it does something that we wouldn't expect with module resolution...

What I wish we could do is:

*   client/tests/simulator_tests/main.rs
*   client/tests/simulator_tests/submodule_a.rs
*   client/tests/simulator_tests/submodule_b.rs

But you can't do that. Instead you have to create a helper module and import that (like you're doing here). This is fine, and allows us to structure and organize our code, but the naming feels a little odd. Naming it `simulator` sounds like it's a module for providing the simulator, not for providing tests relating to the simulator.

---

So there's two options here:

1.  Turn `simulator` into a shared module that provides code for how to interact with the simulator, and then each test is it's own separate integration test (that's a bit annoying, I think; though we can reconsider if we want this in a future PR, we probably shouldn't change to this for now).
2.  Name the module explicitly to suggest it's about organizing test code for `simulator_tests`.

This might be bikeshedding a bit, so feel free to ignore, but WDYT of naming this module `simulator_tests_registration`, to make it clear it's a helper module that is responsible for organizing and registering tests related to `simulator_tests`.

So the final layout would look like this:

*   client/tests/simulator_tests_registration/mod.rs
*   client/tests/simulator_tests_registration/random.rs
*   client/tests/simulator_tests_registration/capability.rs
*   client/tests/simulator_tests.rs

Then in `simulator_tests.rs` we include that module:

```rs
mod simulator_tests_registration;
```

---

This is basically exactly what you're doing here, except that I'm suggesting we rename the `simulator` module to `simulator_tests_registration` to imply a more specific relation to `simulator_tests`.

File: client/tests/simulator/capability.rs [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+
+// This module is added to "mod.rs" to be included in the test. See "mod.rs" for details.
```
Comment (trent-reed) [inline client/tests/simulator/capability.rs:3]: Not sure this is a necessary comment; the comment about test structure and organization in the sibling `mod.rs` is probably sufficient IMHO.

File: client/tests/simulator/random.rs [hunk 31]
```diff
@@ -0,0 +1,98 @@
...
+
+// This module is added to "mod.rs" to be included in the test. See "mod.rs" for details.
```
Comment (trent-reed) [inline client/tests/simulator/random.rs:3]: Not sure this is a necessary comment; the comment about test structure and organization in the sibling `mod.rs` is probably sufficient IMHO.

File: client/tests/simulator/mod.rs [hunk 32]
```diff
@@ -0,0 +1,9 @@
...
+//
+// Add at least one test module file for each chapter in TPM command spec.
+// If there are many tests for a single command, it is recommended to split into command specific files.
+// For example: <chaptername>_<commandname>.rs
```
Comment (trent-reed) [inline client/tests/simulator/mod.rs:7]: My recommendation would be that they create further submodules. E.g.

*   `<chaptername>/mod.rs`
*   `<chaptername>/<commandname_1>.rs`
*   `<chaptername>/<commandname_2>.rs`

Comment (trent-reed) [reply client/tests/simulator/mod.rs:7]: The only place where submodule resolution is weird is directly under the `tests` directory, now that we're under a normal module things should work in a pretty expected manner.

File: client/tests/simulator_tests.rs (hunk 6 above)
Comment (trent-reed) [reply client/tests/simulator_tests.rs]: I left a separate comment here around the oddities of `tests` module resolution. I kind of forgot about that, so I hope it didn't cause too much headaches... Anyways, I think it's worth the headaches, less we eventually have a several thousand-line `simulator_tests.rs` file which will be impossible to maintain...

Thanks for restructuring!

File: client/tests/simulator_tests.rs [hunk 33]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
     assert_eq!(received, expected);
 }
 
+#[test]
+fn test_get_random() {
+    const REQUEST_SIZE: u16 = 32;
+    const MIN_RECEIVED: u16 = 32;
+
+    let (_sim_lifeline, mut tpm) = get_started_tpm();
```
[the remark this answers is not in the document]
Comment (trent-reed) [reply client/tests/simulator_tests.rs]: I opened an issue for this: https://github.com/tpm-rs/tpm-rs/issues/139

File: client/Cargo.toml (hunk 15 above)
Comment (trent-reed) [reply client/Cargo.toml]: Thanks! I think this is the way forward with our tests (adding details to the assertions), so hopefully that's all we need. :)

File: client/tests/simulator_tests.rs (hunk 19 above)
Comment (trent-reed) [reply client/tests/simulator_tests.rs]: It should! Iterators are one of those types that attempt to inline and be written as-if you had written them by hand.

In general, you should feel empowered to use the helper functions of `Option`, `Result`, and `Iterator` as-if you were writing their inner code out by hand; this is very different from what we usually think in C and C++.

My experience has been the compiler is pretty good about this stuff. Most of the time producing something as if you had written the trickier for-loop out yourself. So I think we should trust the compiler for these constructs until we have a situation where we need to be *very certain* about the efficiency of something (and tests I'm not too worried about efficiency, so I'd opt for whatever is most readable here regardless of the cost, personally).

Anyways, this is one of the things that Rust means when it promises these "zero-cost abstractions".

**Comment by trent-reed** [comment]
Okay, next pass done - thanks!

File: client/tests/simulator_tests.rs [hunk 34]
```diff
@@ -83,6 +83,63 @@ fn test_get_capability_manufacturer_id() {
     assert_eq!(received, expected);
 }
 
+#[test]
+fn test_get_random() {
+    const REQUEST_SIZE: u16 = 32;
+    const MIN_RECEIVED: u16 = 32;
+
```
[the remark this answers is not in the document]
Comment (TRJensen) [reply client/tests/simulator_tests.rs]: Fixed

File: client/tests/simulator_tests.rs (hunk 19 above)
Comment (TRJensen) [reply client/tests/simulator_tests.rs]: Thanks for the great clarification 👍 🙂

File: client/tests/simulator/capability.rs (hunk 26 above)
Comment (TRJensen) [reply client/tests/simulator/capability.rs:1]: Fixed

File: client/tests/simulator/mod.rs (hunk 27 above)
Comment (TRJensen) [reply client/tests/simulator/mod.rs:1]: Fixed

File: client/tests/simulator/random.rs (hunk 28 above)
Comment (TRJensen) [reply client/tests/simulator/random.rs:1]: Fixed

File: client/tests/simulator/mod.rs (hunk 29 above)
Comment (TRJensen) [reply client/tests/simulator/mod.rs:3]: Super suggestion - I will adopt 🙂 I don't mind the "bikeshedding" (a word I had to look up, due to English being my 2nd language 😄 ).  It is important to get these details in place from the start, to avoid larger cleanups in the future 👍

File: client/tests/simulator/mod.rs (hunk 32 above)
Comment (TRJensen) [reply client/tests/simulator/mod.rs:7]: I have updated the text recommendation. Fixed.

**Comment by TRJensen** [comment]
@trent-reed @oddcoder Some more adjustments and restructuring based on comments. Please have a final look 👍 🙂

File: client/tests/simulator_tests_registration/mod.rs [hunk 35]
```diff
@@ -0,0 +1,9 @@
+// Please see "simulator_tests.rs" for details.
```
Comment (oddcoder) [inline client/tests/simulator_tests_registration/mod.rs:1]: Preferably use doc comments and not those styles

```
//! Please see ...
```

**Review by oddcoder — APPROVED** [review]
So from my side, I just have one minor suggestion, but aside from that everything else look good.

File: client/tests/simulator_tests_registration/mod.rs (hunk 35 above)
Comment (TRJensen) [reply client/tests/simulator_tests_registration/mod.rs]: Fixed
